design/userdefined-rootdefinehints.md

```diff
@@ -0,0 +1,148 @@
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [References](#references)
+
+<!-- Added by: mikkosest -->
+
+<!--te-->
+
+## Summary
+
+This document explains how rootdevicehints could be given as user defined parameters as part BaremetalHostSpec. The idea is that user would have the possibilty to define selective constraints for root device selection. Sometimes selective constrains will not be completely deterministic, but it is up to the user to decide which hints to use. 
+Originally the only way for an user to select the used root device hints, was to to choose the name of the hardware profile which has hardcoded root device hint values. When exposing root device hints as part of the BMH spec, this will be part of deprecation steps for hardware profile.
+
+## Motivation
+
+We need to be able select root device, when there are multiple devices available so that device is not chosen randomly
+
+### Goals
+
+1. Decide which root device hints are needed be supported.
+2. Agree on the way how information is propagated to ironic through BMO
+
+
+### Non-Goals
+
+The complete set of root device hints as input are not needed, only the relevant ones.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+- BaremetalHostSpec need to be updated to support list of following root device hints:
+* devicename
+* hctl
+* vendor
+* size
+* wwn
+
+Deteriministic selectors. i.e if used together with each other, those should point the same device instance and should not conflict with group selectors below.
+
+Deterministic selectors
+* devicename
+* hctl
+* wwn
+
+Group selectors
+* vendor
+* size
+
+
+The above parameters are optional in BaremetalHostSpec. Originally Hardware profiles contains hardcoded root device hints, which will be now removed. If there is deterministic hint like devicename or wwn used, then those will dominant selectors over other hints which would be ignored. 
+
+The new struct type added to spec could look like below. In the original inventory code there is an Storage struct that has pretty much the same content, but by having a separete struct we can control easily if we want to add or deprecate support for some hints. Size is given in Gigabytes while in Storage it is given in bytes in order to avoid scale typos in size input.
+
+type RootDeviceHints struct {
```

Comment:
What happens if a user specifies

```
rootDeviceHints:
    wwn: abcdefg
    sizeGB: 100
```

and the `abcdefg` device is 50GB?

Or if there is any other combination of multiple values from this struct that don't work in combination?